src/main/java/com/artipie/Pie.java

```diff
@@ -95,12 +99,32 @@ public Response response(final String line, final Iterable<Map.Entry<String, Str
         return new AsyncSlice(
             CompletableFuture.supplyAsync(
                 () -> new Unchecked<>(this.settings::storage).value()
-            )
-                .thenComposeAsync(
-                    storage -> storage.value(new Key.From(String.format("%s.yaml", repo)))
+            ).thenComposeAsync(
+                storage -> CompletableFuture.completedStage(
+                    new RxStorageWrapper(storage).value(
+                        new Key.From(String.format("%s.yaml", repo))
+                    )
                 )
-                .thenApply(content -> new RepoConfig(this.vertx, content))
-                .thenCompose(Pie::sliceForConfig)
+            ).thenCompose(
+                content -> {
+                    CompletionStage<Slice> result;
+                    try {
+                        result = CompletableFuture.completedStage(
+                            new RepoConfig(this.vertx, content.blockingGet())
```

Comment:
@HDouss please don't use `blockingGet` in `Slice` implementations

Reply:
@g4s8  if we want to use a publisher, that's the way to go. Please see the comments history. Do you prefer the (uglier) solution in this commit fe7ef960000d6295fd06307875c5e31cf94f6a1f? Previously we had the content anyway. It is not because we are calling blockingGet that it became synchronous, we are inside a lambda inside a `thenCompose`. This lambda expression is just compiled, it won't get executed until the storage (result of the previous CompletionStage) is ready.

Reply:
@HDouss I see some issues in this PR:
 - redundant usage of `thenCompose`: you are using `thenCompose` with `CompletableFuture.completedStage()` result for each if branch, it can be simplified to `thenApply` and returning just a result without wrapping it with `CompletableFuture`
 - overusing of completable futures in this PRs, same for previous lines where the publisher is wrapped with additional layer of `CompletableFuture`
 - blocking call to future

Instead of that, you have two possible scenarios: check existance before accessing config with `storage.exist()` or handling exception of content reading with proper methods, e.g. using `Flowable.fromPublisher(content)` with mapping to `RepoConfig`

Reply:
@g4s8 I used `thenApply` for storage (see last commit). I cannot use the same because we should pass a `CompeletionStage` in the last stage to `AsyncSlice`. There is no if branches here. If a I test with storage.exist or handle exceptions from `fromPublisher`, how can I compose afterwards? The composition that results in a `CompeletionStage<Slice>` happens on the `CompletableFuture<Single<Content>>` and cannot be done inside the `thenApply` lambda expression.

Reply:
@g4s8 And for the blocking get it is inside a lambda that won't get executed before the previous thenApply completes, it is the same as before, the content is retrieved and then given to the storage, nothing blocking the Slice construction.

Reply:
@g4s8 If you see that we are not understanding each other or that the PR is in a completely wrong way, please assign someone else to the issue

Reply:
@g4s8 Let me summarize my points here:
* we are building an AsyncSlice that should get in its constructor a `CompletionStage<Slice>`, and not a simple Slice, neither a new implementation based on Flowables (or a big redesign will be involved)
* The construction of this `CompletionStage<Slice>` is done "fast" which means that no blocking operation is done before `Pie.response` returns the object
* we cannot check the existence of the repo config when constructing `CompletionStage<Slice>` if we want to chain completables in the construction of the `CompletionStage<Slice>` passed to `AsyncSlice`, neither we could catch proper exceptions (again unless a redesign is involved)

Given the time I took in this issue, and if you don't agree with these points, please assign to someone else, so I don't get penalized too much by refusing the issue. Thanks.

Reply:
@g4s8 ping

Reply:
@g4s8 ping